compiler/src/dotty/tools/dotc/transform/PatternMatcher.scala

```diff
@@ -332,9 +333,19 @@ object PatternMatcher {
           val mt @ MethodType(_) = extractor.tpe.widen
           var unapp = extractor.appliedTo(ref(scrutinee).ensureConforms(mt.paramInfos.head))
           if (implicits.nonEmpty) unapp = unapp.appliedToArgs(implicits)
-          val unappPlan = unapplyPlan(unapp, args)
-          if (scrutinee.info.isNotNull || nonNull(scrutinee)) unappPlan
-          else TestPlan(NonNullTest, scrutinee, tree.pos, unappPlan)
+          val scrutineeTpe = scrutinee.info.widenDealias
+          if (scrutineeTpe.isBottomType || scrutineeTpe.isNullType) {
```

Comment:
I see why `isNullType ` doesn't allow aliases because you follow other methods, but then you need to test those cases. You probably want to add tests such as:
```scala
type X = Null
val Some(_) = (null: X)
```
which (IIUC) will have the same bug?
If it does, this path must be used for any `X <: Null` — a full subtyping test should work, while it seems at least `tp.dealias,symbol eq ...` cannot.

Reply:
Thanks for catching this. Implemented your suggestion.